nx/lib/nx/binary_backend.ex

```diff
@@ -2097,10 +2097,20 @@ defmodule Nx.BinaryBackend do
 
         inverse_permutation =
           permutation
+          |> Enum.filter(&(&1 in Nx.axes(out)))
           |> Enum.with_index()
           |> Enum.sort_by(fn {x, _} -> x end)
           |> Enum.map(fn {_, i} -> i end)
 
+        diff = Nx.rank(out) - length(inverse_permutation)
+
+        inverse_permutation =
+          if diff > 0 do
+            Enum.to_list(0..(diff - 1)) ++ Enum.map(inverse_permutation, &(&1 + diff))
+          else
+            inverse_permutation
+          end
```

Comment:
I'm not _totally_ sure if this is the correct fix. Basically the test for grad(gather) was failing and this fixed it

Reply:
Will we ever hit the else branch?

Reply:
It seems we can. Keeping only the `do` branch works if we change the range to `0..(diff-1)//1` because it seems that for some cases the diff is 0. I think those are cases where the index tensor as many axes as the input tensor.

The `Enum.filter(... in Nx.axes(out))` takes care of the case where the index tensor has fewer axes, and the `do` block takes care of when there are more axes than in the input.